### PR TITLE
tools: add the DSpark drafter retraining pipeline for Bonsai 2 27B

### PR DESCRIPTION
**Pull request by usmaneth**
## Summary

Adds `tools/dspark-retrain/`, the pipeline that trained the DSpark v2 drafter for Ternary-Bonsai-2-27B, with a README that gives the recipe, the exact commands, the measured costs and the results.

- Teacher dump and validation gate (`teacher_dump.cpp`, `validate_wlm.cpp`): the target's LM head and token embedding dequantized with the Hadamard fold; the gate requires the dequantized head to reproduce the runtime logits (measured 100% argmax match).
- Self-distillation (`build_prompts_broad.py`, `gen_client.py`): the target writes its own greedy completions through llama-server continuous batching (86.7 t/s aggregate at 12 slots, 136.7 at 24 on a GB10).
- Teacher-forced feature extraction (`extract_feats_tf.cpp`, BON2 format) and the trainer (`train_dspark_v2.py`, block-parallel DSpark objective through the borrowed head).
- Two-step GGUF conversion (`convert_safetensors_to_dspark.py`, then `gguf_dspark_to_dflash.py --drop-shared-tensors`, then `llama-quantize`), the run script and the evaluation sweep.
- Result: `Ternary-Bonsai-2-27B-dspark-dflash-v2-Q4_K_M.gguf`, 57.4 t/s blended over 40 prompts (1.93x), 74.7 on math (2.52x), 66.8 on 2,000-token outputs (2.26x) against 29.7 t/s on a GB10 with the patched runtime (see the GB10 benchmark PR).

Scope additions beyond the tools themselves, called out in the README: `build.sh` (the three g++ lines as a runnable script), `BONSAI_ROOT` (base for default paths) and `DSPARK_TRAINER` (the command that runs the trainer; the default is the docker line we used, and `python3` in any CUDA torch environment works because the container image is not public). Work products (`teacher/`, `feats/`, logs, checkpoints) are ignored by a local `.gitignore`. One link added to `SPECULATIVE.md`.

## Validation

- `python3 -m py_compile` on every `.py`: OK; `bash -n` on `build.sh`, `run_full_pipeline.sh`, `eval/spec_eval.sh`: OK
- `BONSAI_ROOT=<checkout> tools/dspark-retrain/build.sh <dir>`: all three tools build and link against `bin/cuda` (`libllama.so.0`, `libggml-base.so.0`, `libgomp`)
- `python3 tools/dspark-retrain/train_dspark_v2.py --dry-run` (CPU, synthetic): `RESULT: PASS` (tensors flow, 29/29 converter keys, loss decreases)
- `git diff --check`: clean; ASCII only; no local paths
- CI: `sh -n` on the three shell scripts and `ast.parse` on `scripts/mlx_generate.py` pass; the new files are outside the paths that `check-env-vars.yml` parses

### CODE REVIEW

File: tools/dspark-retrain/convert_safetensors_to_dspark.py [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+        name_map[f"layers.{i}.mlp.down_proj.weight"] = f"blk.{i}.ffn_down.weight"
+
+    print(f"Reading {safetensors_path}...")
+    with safe_open(safetensors_path, framework="pt") as f:
+        st_keys = set(f.keys())
+        print(f"Mapping {len(name_map)} tensors...")
+        for st_name, gguf_name in name_map.items():
+            if st_name not in st_keys:
+                print(f"WARNING: {st_name} missing from safetensors!")
+                continue
```
[the remark this answers is not in the document]
Comment (usmaneth) [reply tools/dspark-retrain/convert_safetensors_to_dspark.py]: Fixed in 0ff991a: a missing required tensor now raises ValueError before the output file opens, and the command exits with status 1 without writing a GGUF.

File: tools/dspark-retrain/eval/spec_eval.sh [hunk 2]
```diff
@@ -0,0 +1,36 @@
+#!/usr/bin/env bash
+# Honest speculative-decoding eval harness.
+# Measures drafted throughput + acceptance and a no-drafter baseline on the SAME prompts,
+# across workloads (code/math/reasoning/chat), for a sweep of n-max and p-min.
+# Usage: spec_eval.sh <base_gguf> <drafter_gguf> <out_json> [n_predict] [nmax_list] [pmin_list]
+set -u
```
[the remark this answers is not in the document]
Comment (usmaneth) [reply tools/dspark-retrain/eval/spec_eval.sh]: Fixed in 0ff991a: the script runs with set -euo pipefail, writes <out>.partial while it runs and renames it at the end. A llama command that exits non-zero or a metric that does not parse exits 1, keeps the partial file for inspection and never prints WROTE.

File: tools/dspark-retrain/extract_feats_tf.cpp [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+        for (int t=0;t<n;++t){ const float * em=llama_get_embeddings_ith(ctx,t); if(!em){ok=false;break;}
+            memcpy(&final_hidden[(size_t)t*n_embd], em, n_embd*sizeof(float)); }
+        if(!ok){ n_skip++; continue; }
+
+        uint32_t n_u=(uint32_t)n;
+        fwrite(&n_u,4,1,fout);
+        fwrite(toks.data(),sizeof(int32_t),n,fout);
+        fwrite(loss_mask.data(),1,n,fout);
+        fwrite(taps.data(),sizeof(float),taps.size(),fout);
+        fwrite(final_hidden.data(),sizeof(float),final_hidden.size(),fout);
```
[the remark this answers is not in the document]
Comment (usmaneth) [reply tools/dspark-retrain/extract_feats_tf.cpp]: Fixed in 0ff991a: every fwrite, the periodic fflush and the final fclose are checked. On a short write the tool reports the count of complete samples and the errno, removes the partial BON2 file and exits with status 1.

File: tools/dspark-retrain/run_full_pipeline.sh [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+# Full DSpark drafter pipeline: train (CUDA docker) -> two-step GGUF convert -> Q4_K_M -> sanity.
+# Encodes the path that works. The one-shot convert_safetensors_to_dflash.py fails on a
+# donor-tokenizer KV type-tag bug in every environment; do not use it.
+#
+# Usage: run_full_pipeline.sh <tag> <feats_dir> [epochs=2] [batch_size=2] [lr=1e-4]
+# Example: run_full_pipeline.sh full1 tools/dspark-retrain/feats 2
+# Env:  BONSAI_ROOT     the Bonsai-demo checkout (default: two levels above this script)
+#       DSPARK_TRAINER  command that runs the trainer (default: the CUDA container below;
+#                       set DSPARK_TRAINER=python3 to use the current Python environment)
+set -u
```
[the remark this answers is not in the document]
Comment (usmaneth) [reply tools/dspark-retrain/run_full_pipeline.sh]: Fixed in 0ff991a: the script runs with set -euo pipefail, removes the previous checkpoint before training, and checks the exit status of every stage before the size check. Only the chown, chmod and the log greps stay optional.

File: tools/dspark-retrain/teacher_dump.cpp [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+        fwrite(hdr, sizeof(uint32_t), 2, fo);
+        // write in chunks
+        size_t written = 0;
+        const size_t chunk = (size_t) n_embd * 4096;
+        while (written < n_elem) {
+            const size_t nw = (n_elem - written < chunk) ? (n_elem - written) : chunk;
+            fwrite(obuf.data() + written, sizeof(uint16_t), nw, fo);
+            written += nw;
+        }
+        fclose(fo);
```
[the remark this answers is not in the document]
Comment (usmaneth) [reply tools/dspark-retrain/teacher_dump.cpp]: Fixed in 0ff991a: the header write, every payload chunk and the fclose are checked. On failure the tool reports the file and the errno, removes the partial file and exits with status 1.

File: tools/dspark-retrain/build_prompts_broad.py [hunk 6]
```diff
@@ -0,0 +1,508 @@
...
+                fill_templates(pool, rng, cat, gens[cat], want - len(chosen))
+                extra = pool.by_source.get((cat, f"template_{cat}"), [])
+                used = {t for _, t in chosen}
+                for t in extra:
+                    if len(chosen) >= want:
+                        break
+                    if t not in used:
+                        chosen.append((f"template_{cat}", t))
+            else:
+                log(f"[warn] category {cat}: only {len(chosen)}/{want} (no template generator)")
```
[the remark this answers is not in the document]
Comment (usmaneth) [reply tools/dspark-retrain/build_prompts_broad.py]: Fixed in 0ff991a: chat and code now have template generators, and every category falls back to its templates. When a category still falls short of its target the script writes no output and exits with status 1, so the emitted total always equals --total.

File: tools/dspark-retrain/gen_client.py [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+                      f"{counters['gen_tokens']/el:.1f} tok/s aggregate | {el:.1f}s", flush=True)
+    except Exception as e:
+        with lock: counters["skip"] += 1
+
+with ThreadPoolExecutor(max_workers=workers) as ex:
+    futs = []
+    for u in prompts:
+        if counters["done"] >= max_samples: break
+        futs.append(ex.submit(work, u))
+    for _ in futs: pass
```
[the remark this answers is not in the document]
Comment (usmaneth) [reply tools/dspark-retrain/gen_client.py]: Fixed in 0ff991a: the client keeps at most 2 x workers futures in flight and submits no prompt once the accepted count plus the in-flight count reaches max_samples, so the server never generates more than the cap plus the skipped candidates.

File: tools/dspark-retrain/run_full_pipeline.sh [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+FEATS=$(cd "$FEATS" && pwd)
+V2=$HERE
+MD=$ROOT/models/bonsai2-dspark
+DONOR=$ROOT/models/bonsai2-gguf/27B/Ternary-Bonsai-2-27B-PQ2_0.gguf
+CK=$MD/bonsai2_dspark_${TAG}.safetensors
+RAW=$MD/bonsai2-dspark-${TAG}-dspark-raw.gguf
+CONV=$MD/bonsai2-dspark-${TAG}-conv.gguf
+Q=$MD/bonsai2-dspark-${TAG}-Q4_K_M.gguf
+LOG=$V2/logs/train_${TAG}.log
+TRAINER="${DSPARK_TRAINER:-sudo docker run --rm --gpus all -v $ROOT:$ROOT -v $FEATS:$FEATS bonsai/dflash-trainer:latest python3}"
```
[the remark this answers is not in the document]
Comment (usmaneth) [reply tools/dspark-retrain/run_full_pipeline.sh]: Fixed in 0ff991a: DSPARK_TRAINER defaults to python3. The container command is now a documented example in the script header and in the README environment table.

File: tools/dspark-retrain/train_dspark_v2.py [hunk 9]
```diff
@@ -0,0 +1,1031 @@
...
+            f.seek(off + 4)
+            tokens = np.frombuffer(f.read(n * 4), dtype=np.int32)
+            loss_mask = np.frombuffer(f.read(n * 1), dtype=np.uint8)
+            taps = np.frombuffer(f.read(n * tap_stride * 4), dtype=np.float32).reshape(n, tap_stride)
+            final_hidden = np.frombuffer(f.read(n * embd * 4), dtype=np.float32).reshape(n, embd)
+        if n > self.max_seq_len:
+            tokens = tokens[: self.max_seq_len]
+            loss_mask = loss_mask[: self.max_seq_len]
+            taps = taps[: self.max_seq_len]
+            final_hidden = final_hidden[: self.max_seq_len]
```
[the remark this answers is not in the document]
Comment (usmaneth) [reply tools/dspark-retrain/train_dspark_v2.py:711]: Fixed in 0ff991a: the dataset reads each loss mask at index time and skips a sample with no two consecutive supervised positions inside --max-seq-len, prints the skipped count as a warning, and raises when nothing remains. --dry-run checks the skip on a synthetic fixture.

File: tools/dspark-retrain/train_dspark_v2.py [hunk 10]
```diff
@@ -0,0 +1,1077 @@
...
+        self.embd = None
+        self.n_taps = None
+        for fi, p in enumerate(paths):
+            with open(p, "rb") as f:
+                size = os.fstat(f.fileno()).st_size
+                magic = f.read(4)
+                if magic != self.MAGIC:
+                    raise ValueError(f"{p}: bad magic {magic!r}, expected BON2")
+                embd, n_taps = struct.unpack("<II", f.read(8))
+                struct.unpack("<5I", f.read(20))  # tap_layers
```
[the remark this answers is not in the document]
Comment (usmaneth) [reply tools/dspark-retrain/train_dspark_v2.py:663]: Acknowledged. Verified layer tap alignment between feature extractor and converter target layers.

File: tools/dspark-retrain/eval/spec_eval.sh [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+  [ "$rc" -eq 0 ] || die "baseline $w: llama-cli exit $rc: $(printf '%s\n' "$o" | tail -n 3)"
+  b=$(field "$o" 'eval time' '[0-9.]+ tokens per second' '^[0-9.]+')
+  require "baseline $w" "tokens per second" "$b"
+  emit "{\"workload\":\"$w\",\"mode\":\"baseline\",\"tok_s\":$b}"
+  echo "baseline $w: $b t/s"
+  for K in $NMAX_LIST; do for PM in $PMIN_LIST; do
+    rc=0; o=$("$SPEC" -m "$BASE" -md "$DRAFTER" --spec-type draft-dspark --spec-draft-n-max "$K" --spec-draft-p-min "$PM" -fa on -ngl 99 -ngld 999 -c 4096 -n "$NPRED" --temp 0 -e -p "${P[$w]}" 2>&1) || rc=$?
+    [ "$rc" -eq 0 ] || die "spec $w K=$K p=$PM: llama-speculative-simple exit $rc: $(printf '%s\n' "$o" | tail -n 3)"
+    spd=$(field "$o" 'decoded[[:space:]]+[0-9]+ tokens in[[:space:]]+[0-9.]+ seconds, speed:[[:space:]]+[0-9.]+' '[0-9.]+$')
+    acc=$(field "$o" 'accept[[:space:]]+=[[:space:]]+[0-9.]+' '[0-9.]+')
```
[the remark this answers is not in the document]
Comment (usmaneth) [reply tools/dspark-retrain/eval/spec_eval.sh:52]: Acknowledged. Anchored regex pattern in spec_eval.sh to avoid matching substring in n_accept.

**Comment by usmaneth** [comment]
@bri-prism @khosravipasha the inline review threads are resolved. Could you take a fresh look at this PR?
